### DISCUSSION TITLE
customChangelogUrl handlebar templating support

### DISCUSSION THREAD

**Discussion by gberche-orange** [renovatebot/renovate#18280] [Suggest an Idea]
I'd like to use a regexp capture group to programmatically construct a changelog url to a set of docker image packages, (such as https://hub.docker.com/r/crossplane/provider-gcp) such as:

```json
    {
      "matchPackagePatterns": ["crossplane/(?<crossplanePackageName>[a-zA-Z0-9\\.\\-]*?)"],
      "customChangelogUrl": "https://github.com/crossplane-contrib/{{crossplanePackageName}}/"
    },
```

However, I see in the following traces logs:

`DEBUG: Fetching changelog: https://github.com/crossplane-contrib/{{crossplanePackageName}}/ (v0.3.0 -> v0.4.1)(branch="renovate/crossplane-provider-kubernetes-0.x")`

This makes me think that customChangelogUrl is not yet subject to templating (as suggested by https://docs.renovatebot.com/templates/#exposed-config-options

Is there another way to avoid duplicating the customChangeLogUrl for each crossplane-package (currently [63 packages](https://github.com/crossplane-contrib?q=provider&type=all&language=&sort=)) with syntax such as the following ?

```json
    {
      "matchPackageNames": ["crossplane/provider-gcp"],
      "customChangelogUrl": "https://github.com/crossplane-contrib/provider-gcp/"
    },
    {
      "matchPackageNames": ["crossplane/provider-aws"],
      "customChangelogUrl": "https://github.com/crossplane-contrib/provider-aws/"
    },
```

Background: 

I have a set of crossplane packages (K8S CR) that reference docker image dependencies who are lacking reference to source repo.

**Comment by viceice** [comment]
you can't use capture groups in an match pattern of package rules. they aren't stored anywhere.

you can use the `replace` template helper to get the name from depName value

**Reply by gberche-orange** [reply]
thanks @viceice , I tried the following syntax, is it what you were suggesting ?

```json
    {
      "description": "crossplane providers docker images are hosted on the crossplane-contrib github org. See discussion about customChangeUrl template at https://github.com/renovatebot/renovate/discussions/18280",
      "matchDatasources": [ "docker"],
      "matchPackagePatterns": [
          "crossplane/(?<crossplanePackageName>[a-zA-Z0-9\\.\\-]*?)"
      ],
      "customChangelogUrl": "https://github.com/crossplane-contrib/{{{replace 'crossplane/' '' depName}}}/"
    },
```

Unfortunately this seems ignored as well from the `customChangelogUrl` rendering according to the following trace. Did I miss something to get the `customChangelogUrl` rendered as a template ?

`DEBUG: Fetching changelog: https://github.com/crossplane-contrib/{{{replace 'crossplane/' '' depName}}}/ (v0.10.0 -> v0.11.1)(branch="renovate/crossplane-provider-helm-0.x")`

**Reply by viceice** [reply]
🤔 are you sure `customChangelogUrl` is a templatable field? it seems it can only be used inside templates.

**Reply by gberche-orange** [reply]
that's what I was originally suspecting in this post https://github.com/renovatebot/renovate/discussions/18280#discussion-4466855

> This makes me think that customChangelogUrl is not yet subject to templating (as suggested by https://docs.renovatebot.com/templates/#exposed-config-options

Would it make sense to make `customChangelogUrl` a templatable field ?

**Reply by viceice** [reply]
i don't think so. you can ask the docker image authors to include labels, so renovate can automatically find the source repos.

we've also a static list included inside renovate for some know packages to their repos

**Reply by gberche-orange** [reply]
thanks @viceice 

I've submitted a suggestion to the crossplane community into https://github.com/crossplane/crossplane/issues/3370 to add the OCI image label. This is likely to take time.

I've currently worked around the issue by duplicating the package rule for each package.

As usage of OCI images are rapidly increasing and the [DockerFile reference](https://docs.docker.com/engine/reference/builder/#label) does not yet make mention of the [OCI label org.opencontainers.image.source](https://github.com/opencontainers/image-spec/blob/main/annotations.md), I believe this is quite common to have OCI images that can benefit from renovate PR but be missing changelog URL in their PRs.

It seems useful to me to provide a way for renovate package consumers to workaround lack of annotations for new communities leveraging OCI images. Make `customChangelogUrl` a templatable field seemed like a possible fix, there are certainly reasons why this does not seem a good idea.

**Reply by viceice** [reply]
it's documented here

https://docs.renovatebot.com/modules/datasource/#docker-datasource
